skxray/tests/test_diff_roi_choice.py

```diff
@@ -0,0 +1,332 @@
+#   the documentation and/or other materials provided with the         #
+#   distribution.                                                      #
+#                                                                      #
+# * Neither the name of the Brookhaven Science Associates, Brookhaven  #
+#   National Laboratory nor the names of its contributors may be used  #
+#   to endorse or promote products derived from this software without  #
+#   specific prior written permission.                                 #
+#                                                                      #
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS  #
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT    #
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS    #
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE       #
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,           #
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES   #
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR   #
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)   #
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,  #
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OTHERWISE) ARISING   #
+# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE   #
+# POSSIBILITY OF SUCH DAMAGE.                                          #
+########################################################################
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import six
+import numpy as np
+import logging
+logger = logging.getLogger(__name__)
+from numpy.testing import (assert_array_equal, assert_array_almost_equal,
+                           assert_almost_equal)
+import sys
+
+from nose.tools import assert_equal, assert_true, raises
+
+import skxray.diff_roi_choice as diff_roi
+
+from skxray.testing.decorators import known_fail_if
+import numpy.testing as npt
+
+
+def test_roi_rectangles():
+    detector_size = (15, 10)
+    num_rois = 3
+    roi_data = np.array(([2, 2, 3, 3], [6, 7, 3, 2], [11, 8, 5, 2]),
+                        dtype=np.int64)
+
+    xy_inds, num_pixels, pixel_list = diff_roi.roi_rectangles(num_rois,
+                                                              roi_data,
+                                                              detector_size)
+
+    xy_inds_m = ([1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3,
+                 3, 3, 3, 3, 3])
+
+    num_pixels_m = [9, 6, 8]
+    pixel_list_m = ([22, 23, 24, 32, 33, 34, 42, 43, 44, 67, 68, 77, 78,
+                     87, 88, 118, 119, 128, 129, 138, 139, 148, 149],)
+
+    assert_array_equal(num_pixels, num_pixels_m)
+    assert_array_equal(xy_inds, np.ravel(xy_inds_m))
+    assert_array_equal(pixel_list, pixel_list_m)
+
+
+def test_roi_rings():
```

Comment:
There is an off-by-one bug in here:

```python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mp

def test_demo(ax, center):
    q_inds_m = np.array([1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 3, 3, 4, 4, 5, 5, 6,
     6, 7, 7, 8, 8, 9, 9, 10, 1, 1, 1, 2, 2, 3, 3, 4, 4,
     5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 1, 1, 1, 2, 2, 3,
     3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 1, 1, 1,
     2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10,
     1, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8,
     9, 9, 10, 1, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7,
     7, 8, 8, 9, 9, 10, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 3,
     3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 2, 1, 1,
     1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7,
     8, 8, 9, 9, 10, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 4,
     4, 5, 5, 6, 6, 6, 7, 7, 8, 8, 9, 9, 10, 2, 2, 2, 2,
     2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8,
     8, 9, 9, 10, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 5,
     5, 6, 6, 6, 7, 7, 8, 8, 9, 9, 10, 10, 3, 3, 3, 3, 3,
     3, 3, 3, 4, 4, 4, 5, 5, 5, 6, 6, 7, 7, 8, 8, 8, 9,
     9, 10, 10, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 6,
     6, 7, 7, 7, 8, 8, 9, 9, 10, 10, 10, 4, 4, 4, 4, 4, 4,
     4, 4, 5, 5, 5, 5, 6, 6, 6, 7, 7, 8, 8, 8, 9, 9, 10,
     10, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 7, 7,
     8, 8, 8, 9, 9, 10, 10, 10, 5, 5, 5, 5, 5, 5, 5, 5, 6,
     6, 6, 6, 7, 7, 7, 7, 8, 8, 9, 9, 9, 10, 10, 6, 6, 6,
     6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 8, 8, 8, 9, 9, 9,
     10, 10, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 7, 8, 8,
     8, 9, 9, 9, 10, 10, 10, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
     7, 8, 8, 8, 8, 9, 9, 9, 10, 10, 10, 7, 7, 7, 7, 7, 7,
     7, 7, 7, 8, 8, 8, 8, 8, 9, 9, 9, 10, 10, 10])

    pixel_list_m = ([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14,
                     15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
                     31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
                     44, 45, 46, 47, 48, 49, 50, 56, 57, 58, 59, 60, 61,
    62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74,
    75, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92,
    93, 94, 95, 96, 97, 98, 99, 100, 106, 107, 108, 109,
    110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120,
    121, 122, 123, 124, 125, 131, 132, 133, 134, 135, 136,
    137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147,
    148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158,
    159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
    170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180,
    181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191,
    192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202,
    203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213,
    214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224,
    225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235,
    236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246,
    247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257,
    258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268,
    269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279,
    280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290,
    291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301,
    302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312,
    313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323,
    324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334,
    335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345,
    346, 347, 348, 350, 351, 352, 353, 354, 355, 356, 357,
    358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368,
    369, 370, 371, 372, 373, 375, 376, 377, 378, 379, 380,
    381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391,
    392, 393, 394, 395, 396, 397, 400, 401, 402, 403, 404,
    405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415,
    416, 417, 418, 419, 420, 421, 425, 426, 427, 428, 429,
    430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440,
    441, 442, 443, 444, 445, 446, 450, 451, 452, 453, 454,
    455, 456, 457, 458, 459, 460, 461, 462, 463, 464, 465,
    466, 467, 468, 469, 470, 475, 476, 477, 478, 479, 480,
    481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491,
    492, 493, 494])

    tt = np.ones((20, 25)).ravel() * np.nan
    tt[pixel_list_m] = q_inds_m
    
    ax.cla()
    ax.set_aspect('equal')
    im = ax.imshow(tt.reshape(20, 25), cmap='Paired', interpolation='nearest')

    [ax.add_artist(mp.Circle(center, radius=j + .5, lw=3, facecolor='none', edgecolor='k')) for j in range(2, 23, 2)]
    Y, X = np.meshgrid(range(20), range(25))
    ax.plot(X.ravel(), Y.ravel(), 'x')
    ax.set_title(str(center))


fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 6), tight_layout=True)
test_demo(ax1, (3, 3))
test_demo(ax2, (4, 4))
```

![so](https://cloud.githubusercontent.com/assets/199813/6137755/7a59ca62-b149-11e4-8b8b-7326ed0df451.png)